chess/position.go

```diff
@@ -0,0 +1,219 @@
+package chess
+
+// A Position weakly describes the state of a chess game.
+type Position struct {
+	Board          Board
+	CastleRights   CastleRights
+	EnPassantRight EnPassantRight
+	ActiveColor    Color
+}
+
+// NewPosition returns a new starting position.
+func NewPosition() Position {
+	return Position{
+		Board:        NewBoard(),
+		CastleRights: WhiteKingSide | WhiteQueenSide | BlackKingSide | BlackQueenSide,
+	}
+}
+
+// Move applies a move to the position. The move must be legal.
+//
+// Move returns true if it resets the FIDE 50-move and 75-move rule
+// counters. See FIDE Laws of Chess, Articles 9.3 and 9.6.2.
+func (p *Position) Move(m Move) (reset bool) {
+	// Determine whether move counters must be reset.
+	fromPiece, _ := p.Board.Get(m.From)
+	_, toPieceExists := p.Board.Get(m.To)
+	if fromPiece.Type == Pawn || toPieceExists {
+		reset = true
+	}
+
+	// Invert the active color.
+	p.ActiveColor = !p.ActiveColor
+
+	// Update the en passant right.
+	p.EnPassantRight.Valid = false
+	if fromPiece.Type == Pawn {
+		fromRank, toRank := m.From.Rank(), m.To.Rank()
+		switch {
+		case fromRank == Rank2 && toRank == Rank4:
+			p.EnPassantRight = EnPassantRight{Square: m.From.Above(), Valid: true}
+		case fromRank == Rank7 && toRank == Rank5:
+			p.EnPassantRight = EnPassantRight{Square: m.From.Below(), Valid: true}
+		}
+	}
+
+	// Clear the "from" square.
+	p.Board.ClearPiece(fromPiece, m.From)
+
+	// Handle castling logic.
+	var castleRightsUsed CastleRights
```

Comment:
nit: I wouldn't call White moving their rook from A1 to A2 "using" a castle right. Better name for variable?

Reply:
`castleRightsToClear`